### ISSUE TITLE
Screen tearing when recursively opening a dictionary.

### ISSUE THREAD

**Issue by Biep** [koreader/koreader#10190]
* KOReader version: v2023-01-18-g0e53631f_2023-02-08
* Device: Kobo Aura One

#### Issue
When closing a recursive dictionary window, intermittently the book text tears over the still-open window - to the left of a straight line from a point roughly halfway at the top border to a point halfway or below at the left border.

#### Steps to reproduce
1. Open a dictionary window (e.g. by long-pressing a word in an epub or pdf).
2.  From this window, open another window (by long-pressing a word in the first window, or by clicking the edit icon at the top right and doing a search from there.  (Repeat ad libitum.)
3. Close the latter window.
4. See the book text, and not the underlying window. to the left of the above-decribed line.  Or not, as the bug is intermittent.

As the "close" button is still visible, on can simply close the lower window and continue.

**Comment by Frenzie** [comment]
That's probably a semi-known issue, though I forget what @poire-z called it and I haven't personally seen it in a few years I don't think.

**Comment by poire-z** [comment]
The unfamous _papercut refresh glitch_ ! #3130 , https://github.com/koreader/koreader/issues?q=papercut
I still see it on my GloHD, but less often than I did 5 years ago.
And I indeed see it most with the DictQuickLookup windows.

**Comment by offset-torque** [comment]
**@ Devs**

I can add this info to the guide so the users would know that:
- this is a known problem with some models
- and it is non-fixable

I had one of these papercut with the orientation change in **Libra**.
Here is **Aura One** and **poire-z** has **GloHD**.
I guess mainly Kobos has this.
Are there other models/brands which has this kind of problem ?

**Comment by poire-z** [comment]
I'm not certain it is non-fixable. It's probably some race condition in the refreshes, that could probably be fixed by delaying some stuff - or having a real look and being sure there are not uneeded consecutive refreshes.
At least, I used to get that often when opening the ToC 5 years ago - but it must have been fixed in passing (while we reworked some stuff, that is: not willingly :) as I haven't seen it for along time.

**Comment by Frenzie** [comment]
It may be in some sense non-fixable from the userland side. Or perhaps rather, fixing it might make the experience minutely worse on devices with better drivers/hardware, but that depends on the exact cause.

@offset-torque 
> I can add this info to the guide so the users would know that:

It's essentially already covered by `Ghosting and residue when turning pages`.

**Comment by Biep** [comment]
It seems to happen more with certain books, e.g.
[The Life and Strange Surprising Adventures - Daniel Defoe.zip](https://github.com/koreader/koreader/files/10941927/The.Life.and.Strange.Surprising.Adventures.-.Daniel.Defoe.zip).
(Stupidly, I cannot drop a .epub here..)

By the way, I just had an interesting experience: a single torn button on the keyboard.  The Delete button was torn, the upper left corner being all grey.

**Comment by NiLuJe** [comment]
By cantfix I mean that it's a race condition (one that is very, very, very, very, very platform specific, in the very largest sense of "platform" i.e., it's temperature sensitive, both ambient and onboard), and all the various things we could do to alleviate it are terrible, terrible, terrible ideas ;).

(The `[ui]` modes *might* help, but depending on what exactly trips it, it could lead to very annoying UX -- more so than the artifact itself ;)).

**Comment by Biep** [comment]
So would serialising not fix it?  Lots of time (relatively speaking) is lost by having the elements of the window (such as its buttons) undo themselves separately - eliminating that would more than make up for serialising the removal of the window (i.e. first repainting the book, and then repainting the underlying window(s) in order).
It is not a usability hindrance at all, merely quite ugly, and possibly scary for a newcomer who doesn't know what is happening.

Serialising could even be an option, so that people who don't experience or care about the problem can go for speed.

**Comment by NiLuJe** [comment]
Err, what do you mean by "serializing", exactly?

**Comment by NiLuJe** [comment]
> By the way, I just had an interesting experience: a single torn button on the keyboard. The Delete button was torn, the upper left corner being all grey.

FWIW, the keyboard was the last place I could somewhat reliably trip this, on the one device where I can actually see this artifact (i.e., the oldest one, the H2O). On newer devices, the highlight tends to be optimized out by the kernel, but without more adverse effects than a missing highlight or two (or, almost all of them if you're on an MTK SoC ;p).

**Comment by NiLuJe** [comment]
> Err, what do you mean by "serializing", exactly?

I wouldn't mind an actual video, that might be quicker than an endless back and forht trying to figure out what we're talking about, exactly ;D.

**Comment by Biep** [comment]
I mean removing parallelism: waiting for the one action to end before starting the next.  I suppose the race condition is between the painting of the background book and the (logically subsequent) painting of the dictionary window.
Not sure how I could video that.  But let me add a photo of the kind of more complex cuts that appear not infrequently.
![Papercut tear](https://user-images.githubusercontent.com/12461597/224573231-8d6c345f-c487-43fc-b592-702db31b6a83.jpeg)

**Comment by NiLuJe** [comment]
As a quick eInk 101 to expand on that:

Drawing to the framebuffer and actually reefreshing the screen are completely separate events on an eInk screen.
We control when both of things happen, and they are completely separate.

Drawing is handled in a bottom-to-top fashion, and is handled and batched *before* (obviously) we even *request* a screen refresh.
A draw request implies (or not, but let's say it does to simplify matters) enquing a matching refresh request for the region repainted (or not, again ;p).

Both of those are queued and batched, so we don't redraw the same thing a billion times; and only once the paint queue has been emptied is the refresh queue processed, so that we only end up requesting a *single* refresh request from the kernel.

And *now* is where things get funky. Notice that I explicitly used the term "refresh request"? That's because the display driver will *also* use a queue to handle collisions in case of successive requests. So a couple of refrehs requests might only lead to a single effective physical refresh. And this is where shit might go weird.
